### ISSUE TITLE
Suddenly app hangs when reading airtable linked fields

### ISSUE THREAD

**Issue by migo4Huuub** [thunkable/thunkable-issues#764]
### Platform
Thunkable X

### Issue/Request Origin
In my App I store all the used data in airtable tables. Concrete example: The main-table is the “tb.ITteam”. In this table I have a couple of usual fields like firstname, lastname, shortcut, etc. And I have some linked fields like function and group. Function is linked to table “tb.function”, group is linked to table “tb.group”. I use this method to ensure, that values in the fields function and group (in table tb.ITteam) are crosschecked for validity in tb.function and tb.group. Everything worked fine for about two weeks. In the block-snip you can see how I read those fields:

  STACK TRACE AND/OR SCREENSHOTS
![Snap_2020 12 03_20h43m31s_001_](https://user-images.githubusercontent.com/72560309/101208425-76121400-3672-11eb-8113-c20c008898c5.jpg)

You can see (No. 1 & 2): these are the linked fields, all others are unlinked fields.

Without me changing anything (neither in the app, nor in the tables) suddenly, my app hangs when reading 1 & 2. I am sure about the source of the problem, because a ran a test where I deactivated only the blocks 1 & 2 and everything ran fine (of course - without the values function and group). This happens on all platforms: on phones with iOS & Android, on web & in live-tests in the thunkable app. I'm right in the middle of a beta-test currently and now I have around 20 users with a suddenly non-working app, which is pretty unluckily. Any ideas?

**Comment by camtez** [comment]
I'm also experiencing the same issue! Pretty sure this is affecting the getCell block only for me and only on downloaded apps. This is killing me!

**Comment by migo4Huuub** [comment]
Unfortunately there is still no solution on this issue. I also contacted airtable-support, but they requested me to talk to thunkable support. Meanwhile I got some answers from thunkable-support, but all are like "we will look into this" and "please try it now again, we changed something". As a workaround I meanwhile cleared the values in all linked fields, which is working, but of course, annoying (btw: I also removed the fields completely already and recreated them. As soon as I fill them with values all user-apps stop working. To me this is a very serious issue, because nobody can explain, what exactly makes my app stop working, while it is already rolled out to users. If I cannot rely on such a basic functionality (or at least on a support who is able to solve this within an acceptable time) I have to reconsider whether thunkable is my future development platform. I think you can read my disappointment.

**Comment by MrMazzone** [comment]
Hmmm... This is probably the issue with my apps using Airtable.  Following and hoping for a fix.

**Comment by josmas** [comment]
All, is this still an issue?
cc/ @StevenPurn

**Comment by migo4Huuub** [comment]
Yes, it is. If I can assist with any troubleshooting attempts, I'd be happy to do so!

**Comment by StevenPurn** [comment]
Thanks for tagging me. I hadn't seen this before now but I'll be sure to take a look. I'll report back once I have an idea as to what is going on.

**Comment by StevenPurn** [comment]
@migo4Huuub are you able to share your app (or a simplified version) demonstrating the issue? I tried to recreate it with a simple example pictured below but it is working for me. So to resolve, I need to determine if the issue is actually fetching the data or if it is not accepting the id you're passing in correctly. How are you storing the ids to access later? It is possible something broke and they aren't being stored in the Group or Funktion object correctly any more.

![Screen Shot 2020-12-21 at 1 53 14 PM](https://user-images.githubusercontent.com/8118591/102825721-2c286e00-4394-11eb-9df1-d29245d19963.png)

**Comment by migo4Huuub** [comment]
Hello Steven, of course - with pleasure! I just made a copy of my app as a debugging version that you can use for this. You just have to tell me which way I can provide it to you.

Since the app is already in use by users, I have anyway already integrated a function with which I can switch back and forth between real and test environment at the push of a button, so that I do not disturb the real operation during my tests and further developments.

In the debugging version I can provide you now, this function is limited to the test system. In this case the "test" tables are read in the airtable DB instead of the "real" tables (same name, supplemented with a **Q** at the beginning). However, the error we are looking for occurs in the same way in the test environment.

At the moment, the "Function" and "Group" fields are cleared so that the app doesn't hang. As soon as I fill them with values again, the error is immediately there again and reproducible at any time.

I have also created an editor access for you in the Airtable DB, to which I can send you a link (of course not here via the forum, but somehow directly, please tell me how). 

To the explanation: To my app belong currently four tables.
In the **real** system (please **don't change** any data) these are:
tb.itteam, tb.function, tb.group, tb.emergency. 

In the **test system** (that's what you can use), these are:
**q**tb.itteam, **q**tb.function, **q**tb.group, **q**tb.emergency (additionally there is qtb.location. This is part of the current development for the upcoming version and is not part of our bug review here).

I would give you the credentials in the app for my own account and explain with some screenshots how to reproduce the error.
By what channel can we share the data without it being publicly read here?

Thanks and greetings
Migo

**Comment by StevenPurn** [comment]
It doesn't have to be the exact app you're using. If you can make a small sample app with a similar set up or block flow that exhibits the same behavior that would be great. I just need a way to recreate what is happening so I can find the root cause and get to fixing it. If you do want to share the original app, you can send a link to me at steven@thunkable.com.

**Comment by migo4Huuub** [comment]
Okay here we go :)
I prepared a small sample app which I will share to you, @StevenPurn. Additionally you will get an editor access to the "MigoBugTest" Airtable-DB, which I just created for this purpose. To all the other users, who are interested in this issue, I will try to describe here what I did and how you can reproduce the faulty behavoir.

Just as a reminder and to bring all on the same page here: The way of how I read the values from the airtable-DB worked fine for a couple of weeks. I even have distributed apps to Google-Playstore and to Apple-Appstore with users who are involved in my betaphase. Without me changing anything (neither in the app, nor in the airtable-DB), suddenly the app hang while reading from the database. I figured out that the trigger of this hanging is the reading of linked fields in the airtable-DB. When I clear the fields, the app continues working. As soon as I refill the linked fields with values, the app hangs.

Here comes my test-environment:
![grafik](https://user-images.githubusercontent.com/72560309/103091096-56c43200-45f3-11eb-9d4a-7e7b17fbdb7d.png)
I have three records in my main-table "tb.itteam". Beside "Name" I have two additional fields: One is "Country", which is just usual textfield and I have "Function", which is a linked field to table tb.function:
![grafik](https://user-images.githubusercontent.com/72560309/103091212-bf131380-45f3-11eb-9691-28ebccea9c8f.png)
You can see: Record "Hugo" is not linked, cause the field is still empty. Record "Migo" is linked to "Developer". Record "Steven" is linked to "Supporter".

And this is how my test-app works:
![grafik](https://user-images.githubusercontent.com/72560309/103091451-84f64180-45f4-11eb-9e8f-5461dfd3bd0e.png)
1 = Text_Input Field to enter the name
2 = Button to start the read of the DB
3 = Label 1 to display the content of "Country"
4 = Label 2, to display the content of "Function"
5 = Label 3 for control-purposes. As soon as the Read-Button is pushed, it will display "working ...". As soon as the routine finished, it will display "Finished :)".

Here are the blocks:
![grafik](https://user-images.githubusercontent.com/72560309/103091701-4f058d00-45f5-11eb-87c6-34e1330bf5a6.png)
1 = Initialize Label 3 with "working ..."
2 = Look in Airtable-DB "itteam" for the record which matches to the "Name" from the input-field
3 = Wait for DB (airtable access works asynchronously)
4 = Display content of "Country" in Label 1
5 = Display content of linked field (!) "Function" in Label 2
6 = Report "Finished :)" in Label 3.
 

Run without the failure:
If I start the app and enter the name "Hugo" (which is not linked, cause I cleared the value):
![grafik](https://user-images.githubusercontent.com/72560309/103092072-5b3e1a00-45f6-11eb-8151-2d9498b3ecee.png)
1 = Entered name is Hugo
2 = Label 1: Correct Country "IN" is displayed
3 = Label 2: Stays empty, which is correct, cause the field value is cleared
4 = Label 3: Reports "Finished :)" 


Reproduce the failure:
If I start the app and enter the name "Migo" (which is linked to "Developer"):
![grafik](https://user-images.githubusercontent.com/72560309/103092219-d7386200-45f6-11eb-9242-cc949d167f1f.png)
1 = Entered name is Migo
2 = Label 1: Correct Country "DE" is displayed
3 = Label 2: Stays empty allthough the field is filled with "Developer"
4 = Label 3: Still reports "working ..." - the app hangs on reading the linked field above and does not come to an end


Cross-Check: I filled the "Function"-filed in record "Hugo" with "Developer" and run the same test. App hangs on reading. You can easily reproduce it by filling and clearing the linked field back and forth - the misbehavior comes and goes with the filled/cleared field.

PS: Merry christmas, all :)!

**Comment by migo4Huuub** [comment]
Short addition: I asked myself whether this behavior would change, when I access the airtable-DB via API and handle the data-exchange as JSON. But the **result is the same**. Here is the JSON-version of my test:
![grafik](https://user-images.githubusercontent.com/72560309/103151257-b69a1480-477c-11eb-8d94-965eb35934db.png)

**Comment by StevenPurn** [comment]
Great, thanks for sharing! I'll take a look and see if we can figure out how to get this working for you.

**Comment by StevenPurn** [comment]
@migo4Huuub Thanks again for sharing your findings! It always makes it easier to determine what the issue might be. I've been looking into this and I'm curious if you could try something for me to attempt to confirm a suspicion I have. It seems like the Airtable API is returning the linked values as arrays so the block is passing the array as the id rather than the actual id value. Our system is unsure how to handle it so it seems like there is an error occurring somewhere along the way. It seems odd that it used to work for you unless Airtable changed their API recently but that seems unlikely. In any case, can you give this a shot and let me know?

Towards the bottom of your screenshot where you have the purple `get value from` block, please add a block to get the first item from the list of the object's property. Let me know if that helps to resolve the issue.  It would end up looking something like this:
![Screen Shot 2021-01-05 at 3 37 46 PM](https://user-images.githubusercontent.com/8118591/103711317-02627000-4f6c-11eb-928b-16b71b3f8f3f.png)

Another thing I noticed in this specific example is that you seem to be using the variable from the `for each item` block outside of the scope of the block. Generally this is not recommended as it can cause issues accessing the value of the variable. If you want to use the record where the variable matches the text input you should complete that work inside the loop or otherwise store the data you need to use again after the loop is completed. I don't think this is the best explanation so I can expand on it if you need but just wanted to point that out while I was taking a look.

**Comment by migo4Huuub** [comment]
Thanks so much for your investigation, @StevenPurn!

Unfortunately, my tests did not deliver the desired result - only a small improvement: the app no longer gets stuck while reading ;). But it does not grab the correct value.

I first took your hint with the array during testing and implemented it. As result in label 2 only "null" is returned.

Then I added your second hint to the test and cached the variable from the "for each item" block in another object. Unfortunately, nothing changes in the result, the display in label 2 (as soon as a value is filled) is "zero". Here are my blocks: 

![image](https://user-images.githubusercontent.com/72560309/103816144-cf4fd780-5064-11eb-8b53-35bea305c3a8.png)

This is how the test looks like:
![image](https://user-images.githubusercontent.com/72560309/103816499-5b61ff00-5065-11eb-9c67-2b636ab9c5ae.png)


I am happy to stand by if there are further tests to be done. Do you have any other ideas?
Thanks
Migo

**Comment by migo4Huuub** [comment]
Hello @StevenPurn , hope you are fine.

Did you have the possibility to have another look into this topic? Any other ideas? Any things you want me to test?
Thanks, Migo

**Comment by StevenPurn** [comment]
Hi @migo4Huuub! Thanks for following up. I'm doing some more digging on this. I think I might have slightly misunderstood what your blocks were intended to do so can you try reverting it back so the row id that you pass in is just the `get property Function from object app variable zw.record` rather than using the get first item from list block? Let me know if that works.

**Comment by StevenPurn** [comment]
I understand why you think it might be an issue with the get value block. However, I'm able to get it working (and was unable to reproduce the error until now). 

An important thing to note when you're reporting a bug is that distilling it down to the most simple representation is the best way to report it. There was some confusion on this thread because your example was a little more complicated than it needed to be (which is totally fine 😃 , I'm just trying to help explain because this will also help you debug your own app while you build it) which meant I had to do a bit more digging to find the root of the problem. Also, you have blocks that are calling functions which aren't included in the screen shot so I'm not sure what those are doing. 

In actuality, there is no issue with the get field block that I can find. I believe the issue here is within the ordering of your blocks. I noticed you have a function to `WaitForAirtable` so you have kind of already identified the issue. I would love to know what you're doing in that function to see if we should expect these blocks to work as written but here is my working version of your code (slightly modified due to differences in Airtable bases) and my explanation as to what I think is going wrong in your blocks. 

First, here is a picture of my Airtable base for reference:
![Screen Shot 2021-01-19 at 9 56 03 AM](https://user-images.githubusercontent.com/8118591/105073929-963c3d80-5a3c-11eb-8fe7-3bdcccafb3d7.png)

Second, here is a photo of all of the blocks in my project:
![Screen Shot 2021-01-19 at 9 55 37 AM](https://user-images.githubusercontent.com/8118591/105073879-81f84080-5a3c-11eb-8c15-689723d44df9.png)

The key difference here is the placement of the blocks to set the text values of the various components. What I think is happening is that the logic you've implemented to wait for Airtable is not waiting long enough for the values to actually come back. As you seem to be aware since you've added logic to wait, the blocks after the `GetAllRows` block are not guaranteed to run after the values come back from the API call because that block does not stop further execution. But the block to get the field value _is_ blocking execution. So, likely, that block is running before the value is populated, but since it is also a network call, it takes long enough that the value is populated by the time it gets to the final block to set the text to the record value. Does that make sense? If I'm right, you should be able to easily test the order of execution by changing something in both locations and noting which runs first. Here is an example of what that might look like if you use the background color of the screen as your indicator. If at the end of the blocks running, the screen is black it means that the blocks to set the text in the components are running first and so the variable's value isn't populated yet. If the screen's background is red then everything should work since the loop finished before we attempt to use the value we set inside of the loop. 

![Screen Shot 2021-01-19 at 10 01 57 AM](https://user-images.githubusercontent.com/8118591/105074762-aacd0580-5a3d-11eb-8e00-6d5256952bf0.png)

Can you show me what you're doing to WaitForAirtable? Can you also try out this suggestion and see if it fixes the issue for you? Let me know if you have any more questions.

**Comment by migo4Huuub** [comment]
Thank you @StevenPurn so much for this detailed feedback! I will start at the point and test. But first here is what you wanted to know - my Wait4Airtable routine. I had quickly discovered in my first attempts that I need to adjust to asynchronous processing. In the beginning, I had created fixed wait blocks (with static 1 or 2 seconds). Sometimes this worked, sometimes not, because too short. That's why I built this event-triggered wait loop which always works fine. This is what it looks like:  

![image](https://user-images.githubusercontent.com/72560309/105079270-784bf880-5a8f-11eb-9f37-2c83466b3900.png)

**Comment by migo4Huuub** [comment]
@StevenPurn, I have really good news: I can present a solution that works. And now you have to find out why it no longer works the way I did it before 😉:

This is the "get value" as it used to work in the past and now causes the hang (you know it already 😉:

![image](https://user-images.githubusercontent.com/72560309/105195390-e13c7a80-5b3a-11eb-9ac5-516e26cad30e.png)


And this is what I figured out: If I pass the value not directly, but thru a join-function, it works. My understanding is that it shouldn't make any difference at all - and that backs up my suspicion that we're looking for a bug/change in API behaviour here.

![image](https://user-images.githubusercontent.com/72560309/105195781-3f695d80-5b3b-11eb-86e4-be0bc638fee4.png)

What do you think?

**Comment by StevenPurn** [comment]
Thanks for following up. The issue here is still that Airtable responds with an array when the column type is `Link to Table` even if there is only one value because you can add multiple values:
![Screen Shot 2021-01-20 at 9 26 06 AM](https://user-images.githubusercontent.com/8118591/105211801-8a667f00-5b01-11eb-85cb-ec5893f3b992.png)

The reason that the join block works is because it transpiles the resulting to code to:
```
code: String(['test', 'again'])
result: "test,again"
code: String(['test'])
result: "test"
```

So I would guess your code would also work if you took the first value in the list of the property `Function` in your variable. However, you said it wasn't working previously with this change which is very confusing to me since it should effectively be the same result as using the join block in this scenario. 

In the case where the value of the property in question is an array (list) with just one item, converting it to a string (text) directly yields the same result as getting the first value in the list:

![Screen Shot 2021-01-20 at 9 35 54 AM](https://user-images.githubusercontent.com/8118591/105213075-1331ea80-5b03-11eb-96e0-3216c002e946.png)

However, if there is two items in the list, Button 2's text will end up being `hello,world` whereas Button 3's text will just be `hello`:

![Screen Shot 2021-01-20 at 9 37 42 AM](https://user-images.githubusercontent.com/8118591/105213235-4ffde180-5b03-11eb-87a9-8f5db6abb30c.png)


So, if this worked before I'm guessing it was for one of two reasons:
1. You were not previously using this same code with a column which was a `Link to Table` type (and therefore not an array).
2. Airtable recently changed the way their API and schema is set up to make the `Link to Table` type an array when it previously wasn't.

I think the second scenario is less likely but definitely possible. 

It makes sense that it wasn't working before because instead of passing in the record id as a string, it was actually being passed in as an array with one value inside which was the record id in string form. The get value block does not expect an array and so it could not get the value.

Do you have any thoughts on why it wasn't working for you when attempting to get the first item in the list rather than using the join block? Either is a valid solution (assuming you don't ever add more than one value in a row in that column) so I would consider this working at the moment. A way to confirm would be to get the record id, insert it into a new column which is just text and try to recreate the same code using that column. If it works, it means there isn't an issue with the get value block and the issue is with the input. 

Let me know if you have any other questions.

**Comment by migo4Huuub** [comment]
Thanks again @StevenPurn for your detailed explanations about the array in the linked fields. I have rebuilt my test scenario to test again your "in list get #1"-suggestion and can report back: You are right. The LIST-approach actually works just like the JOIN-approach. I probably overlooked an error during the first test, I'm sorry.

Which means:
![image](https://user-images.githubusercontent.com/72560309/105234590-cc2c0f80-5b6b-11eb-84e1-56479ca53c3e.png)
creates the same (correct) result as:
![image](https://user-images.githubusercontent.com/72560309/105234716-ef56bf00-5b6b-11eb-8bfc-4abeeeb437e2.png)



Regarding the question "what was the reason? / what was changed?": due to the fact that this problem suddenly appeared - and also with apps that I already have live in the Appstore and Playstore, it must be the Airtable API in my view. I haven't put any new releases in the stores since the error occurred. The apps in use today are only running with my users because I have cleared the Airtable linked fields. I checked it a few minutes ago to be sure: as soon as I fill values into the fields, the users' apps hang when reading. And this unchanged version installed on the devices delivered correct values til some point in december.

I can now finally (thanks so much, Steven!) get new versions of the apps into the stores in which I adjust to the array-linked fields.

Take care!
Migo
